src/Common/MemoryTrackerSwitcher.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <Common/CurrentThread.h>
+#include <Common/MemoryTracker.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+struct MemoryTrackerSwitcher
+{
+    explicit MemoryTrackerSwitcher(MemoryTracker * new_tracker)
+    {
+        if (!current_thread)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "current_thread is not initialized");
```

Comment:
This breaks `clickhouse-disks` for S3 disks

```
Failed to make request to: http://localhost:11111/test?list-type=2&max-keys=1&prefix=clickhouse-disks%2Fdefault%2Ftest.copy: Code: 49. DB::Exception: current_thread is not initialized. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):
```

So after this change it is not possible to do some operations that requires pools from the main thread.

I will fix it here - [`b2ea45b` (#51448)](https://github.com/ClickHouse/ClickHouse/pull/51448/commits/b2ea45b5360972ff609fb43ebf87cb1f99cdcae1)